desmod/workspacesync/s3.py

```diff
@@ -0,0 +1,59 @@
+from concurrent.futures import ThreadPoolExecutor
+
+import boto3
+import getpass
+import os
+
+
+class S3Sync(object):
+    """S3 workspace synchronization.
+
+    The :class:`S3Sync` class implements a sync method that synchronizes the
+    workspace artifacts to an S3 bucket with a configured key prefix.  The
+    s3 destination is of the format: /{prefix}/{workspace name}/{artifact path}
+
+    :param dict config: A fully-initialized configuration dictionary.
+    :param list artifacts: A list of artifact paths relative to the workspace
+        directory. E.g., ["./0/results.yaml"].
+
+    """
+
+    MAX_THREADS = 10
+
+    def __init__(self, config, artifacts):
+        self.client = boto3.client('s3')
+        self.config = config
+        self.workspace = config['sim.workspace']
+        self.artifacts = artifacts
+
+    def _upload_artifact(self, artifact):
+        dest = os.path.join(
+                self.config['sim.workspace.s3_sync_prefix'],
+                os.path.split(self.workspace)[1],
+                (artifact[2:]))
+        self.client.upload_file(
+                artifact, self.config['sim.workspace.s3_sync_bucket'], dest)
```

Comment:
We should extract the required configuration prior to submitting to the thread pool. The config is user input and thus may have all manner of garbage in it.

A good case for retaining the class-oriented strategy here would be so that we can extract configuration in `__init__()` before simulation and then have confidence that the user provided valid configuration well before end of simulation.

    class _Workspace(...):
        def __init__(self, ...):
            ...
            if config.setdefault('sim.sync.s3.enable', False):
                self.s3_sync = S3Sync(...)
            else:
                self.s3_sync = None
            ...
        def __exit__(self, ...):
            ...
            if self.s3_sync:
                self.s3_sync.sync()

    class S3Sync(...):
        def __init__(self, config, ...):
            self.bucket = config['sim.sync.s3.bucket']  # user must provide bucket
            self.prefix = config.setdefault('sim.sync.s3.prefix', '/')
            ...